### ISSUE TITLE
At high CPU load, tokio::time::interval can trigger twice

### ISSUE THREAD

**Issue by sdfgeoff** [tokio-rs/tokio#2301]
**_I worked through this as I wrote the issue, so the likely underlying cause is only mentioned near the end._**

## Version

└── tokio v0.2.13
    └── tokio-macros v0.2.5

## Platform

Linux tabtop2 5.5.6-arch1-1 #1 SMP PREEMPT Mon, 24 Feb 2020 12:20:16 +0000 x86_64 GNU/Linux

(Archlinux)

## Description
I am currently evaluating Tokio as a scheduler for the logic management in a computer game. As a result, I'm very interested in it's performance as it has to meet several soft-real-time targets. As it's a game, most logic has to run repeatedly, so it would be built on top of tokio::time::interval.

I started off by benchmarking: Each task records the time when it actually executes - which allows examination of how well the tokio scheduler is working.

The inside of a task is:
```rust
	async fn run(mut self) -> Self {
		let mut delay = tokio::time::interval(self.run_interval);
		delay.tick().await; // Defaults to being ready
		
		self.start_time = Some(Instant::now());
		let end_time = self.start_time.unwrap() + self.run_duration;
				
		loop {
			delay.tick().await;
			let cur_time = Instant::now();
			self.run_times.push(cur_time);
			if cur_time > end_time {
				break;
			}
		}
		self
	}
```

I then plotted the error - the deviation between when a task expected to be run and when it actually ran. In an ideal world, this histogram would be a spike at zero (Every task ran exactly when it was scheduled to). In a non-ideal world, I would expect it to form either a gaussian distribution around the scheduled time (some just before, some ran just after), or perhaps a half-gaussian distribution such that they only run after their scheduled time.

Running with 400 tasks scheduled from 2ms - 40ms intervals produced the expected distribution:
![out](https://user-images.githubusercontent.com/13490050/76046575-1f818780-5fc5-11ea-9112-8acb795d540d.png)
Horiozontal scale is the "error" from when the task was expected to be run measured in us. In this case, all the tasks ran within 2.5ms of their target time.

However, when the number of tasks was increased to 2000 (which uses about 50% of all 8 threads on my computer), the graph does something strange:
![out](https://user-images.githubusercontent.com/13490050/76046726-93bc2b00-5fc5-11ea-9f87-c67d9806a3ac.png)

Predictably, it's a lot more spread out - the CPU load is higher, so some tasks have to be deferred longer (and somehow, tokio is running some tasks early as well). However, many of the tasks run exact multiples of 1000us before their scheduled time.

By running a bunch of tasks with the same interval, it can be seen that a task has a probability of running exactly one period before it is intended to. 1000 tasks running every 4ms produces (again, to put load on the CPU):
![out](https://user-images.githubusercontent.com/13490050/76047418-76885c00-5fc7-11ea-8bfe-556680621007.png)

------------------------------

By looking at the raw intervals rather than the error, the reason can be deduced:
![out](https://user-images.githubusercontent.com/13490050/76055014-58c2f300-5fd7-11ea-8204-9789af1b5647.png)
All the tasks have some probability of running twice in a row (close to zero time between them). The shorter the task, the higher probability this has of occurring.

--------------------------------
I'm not experienced in async programming, so it may be that 
```
loop {
	delay.tick().await;
	do_stuff()
}
```
is not "the proper way". If so, please let me know!

---------------------------------

In summary:
It appears that awaiting tokio::time::interval can result in tasks running twice. For whatever reason this problem is only evident at higher CPU loading (or with high numbers of tasks, I cannot think of a way to separate these two causes).

**Comment by goffrie** [comment]
This is probably related to the fact that tokio's timer implementation only tracks time down to a granularity of 1 millisecond. If the last wakeup was delayed enough that the next one would be in the "current" millisecond, the `.await` probably completes synchronously (without returning to the task scheduler), causing a zero reading.

**Comment by luksan** [comment]
Could this be fixed by checking if the computed now on line 131 below is earlier than Instant::now(), and in that case recompute now as Instant::now() + self.period?

https://github.com/tokio-rs/tokio/blob/ba8680d66773677f201a26b756e41e9a16b458fe/tokio/src/time/interval.rs#L124-L140

**Comment by barafael** [comment]
Is this solved by https://docs.rs/tokio/latest/tokio/time/enum.MissedTickBehavior.html ?
